jdbc-driver/src/main/java/com/google/cloud/dataproc/jdbc/DataprocInfo.java

```diff
@@ -0,0 +1,169 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.dataproc.jdbc;
+
+import static com.google.cloud.dataproc.jdbc.HiveUrlUtils.checkUrl;
+
+import com.google.api.gax.rpc.ApiException;
+import com.google.cloud.dataproc.v1beta2.Cluster;
+import com.google.cloud.dataproc.v1beta2.ClusterControllerClient;
+import java.io.IOException;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/** Helper class to get cluster info through Dataproc API. */
+public class DataprocInfo {
+
+    private static final String HIVE_PROTOCOL = "jdbc:hive2://";
+    private final ClusterControllerClient clusterControllerClient;
+    private final HiveJdbcConnectionOptions params;
+
+    // Constructor dependency injection
+    public DataprocInfo(HiveJdbcConnectionOptions params, ClusterControllerClient controller) {
+        this.params = params;
+        this.clusterControllerClient = controller;
+    }
+
+    /**
+     * Format client passed in url that is accepted by Dataproc to be Hive acceptable format.
+     * jdbc:hive2://<host>:<port>/<dbName>;transportMode=http;httpPath=<http_endpoint>;<otherSessionConfs>?<hiveConfs>#<hiveVars>
+     *
+     * @return
+     * @throws InvalidURLException
+     */
+    public String toHiveJdbcUrl() throws InvalidURLException {
+        String hiveJdbcURL =
+                String.format(
+                        "%s%s:%d/%s;transportMode=%s;httpPath=%s",
+                        HIVE_PROTOCOL,
+                        getHost(),
+                        params.port(),
+                        params.dbName(),
+                        params.transportMode(),
+                        params.httpPath());
+        if (params.otherSessionConfs() != null) {
+            hiveJdbcURL = String.format("%s;%s", hiveJdbcURL, params.otherSessionConfs());
+        }
+        if (params.hiveConfs() != null) {
+            hiveJdbcURL = String.format("%s?%s", hiveJdbcURL, params.hiveConfs());
+        }
+        if (params.hiveVars() != null) {
+            hiveJdbcURL = String.format("%s#%s", hiveJdbcURL, params.hiveVars());
+        }
+        return hiveJdbcURL;
+    }
+
+    /**
+     * Use Dataproc client libraries to get cluster.
+     *
+     * @return name of the cluster's master node
+     * @throws InvalidURLException
+     */
+    public String getHost() throws InvalidURLException {
+        String host;
+        if (params.clusterName() != null) {
+            host = getClusterByName();
+        } else {
+            // clusterPoolLabel can be null, but filter is not empty string since
+            // it at least has "status.state = RUNNING"
+            String filter = formatClusterFilterString();
+            host = findClusterInPool(filter);
+        }
+        return host + "-m";
```

Comment:
Can be replaced with ternary operator:
```java
String host =
    params.clusterName() == null
        // clusterPoolLabel can be null, but filter is not empty string since
        // it at least has "status.state = RUNNING"
        ? findClusterInPool(formatClusterFilterString());
        : getClusterByName();
```